### PR TITLE
ceph_mon: Fix MonitorDBStore usage

### PR DESCRIPTION
**Pull request by Matan-B**
Fixes: https://tracker.ceph.com/issues/61820

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mon/MonitorDBStore.h [hunk 1]
```diff
@@ -709,7 +709,7 @@ class MonitorDBStore
   }
 
   void close() {
-    // there should be no work queued!
+    io_work.wait_for_empty();
```
Comment (batrick) [inline src/mon/MonitorDBStore.h:712]: Considering the comment says `there should be no work queued!` quite explicitly, I would expect a new comment explaining why we are now waiting for it to empty.

**Review by batrick — CHANGES_REQUESTED** [review]
Rest looks fine to me.

File: src/mon/MonitorDBStore.h (hunk 1 above)
Comment (Matan-B) [reply src/mon/MonitorDBStore.h]: After increasing finisher's debug level and found out that the queue was indeed empty when closing.
```
2023-07-20T09:18:50.796+0000 7fcf5c740cc0 10 finisher(0x5570504d85a0) wait_for_empty empty
2023-07-20T09:18:50.796+0000 7fcf5c740cc0 10 finisher(0x5570504d85a0) stop
2023-07-20T09:18:50.796+0000 7fcf467ef700 10 finisher(0x5570504d85a0) finisher_thread stop
2023-07-20T09:18:50.796+0000 7fcf5c740cc0 10 finisher(0x5570504d85a0) stop finish
2023-07-20T09:18:50.796+0000 7fcf5c740cc0  4 rocksdb: [db/db_impl/db_impl.cc:496] Shutdown: canceling all background work
2023-07-20T09:18:50.796+0000 7fcf5c740cc0  4 rocksdb: [db/db_impl/db_impl.cc:704] Shutdown complete
2023-07-20T09:18:50.796+0000 7fcf5c740cc0  0 ceph-mon: created monfs at /var/lib/ceph/mon/ceph-a for mon.a
```
I added an assertion instead (38b94fba937e54a9ccdc435e1903ee785a50a9ab).

**Comment by Matan-B** [comment]
The reported jobs before the changes here:
**dbench:** https://pulpito.ceph.com/matan-2023-07-13_12:15:57-fs:verify-reef-distro-default-smithi
**fsstress:** https://pulpito.ceph.com/matan-2023-07-13_12:15:14-fs:verify-reef-distro-default-smithi

After:
**dbench:** https://pulpito.ceph.com/matan-2023-07-20_07:03:54-fs:verify-wip-matanb-reef-61820-v2-distro-default-smithi
**fsstress:** https://pulpito.ceph.com/matan-2023-07-20_07:03:15-fs:verify-wip-matanb-reef-61820-v2-distro-default-smithi

Since the issue appears in R only, I think we can test it in main first and later on apply this to R if all looks good.

File: src/ceph_mon.cc [hunk 2]
```diff
@@ -356,6 +356,8 @@ int main(int argc, const char **argv)
     exit(1);
   }
 
+  auto store = std::make_shared<MonitorDBStore>(g_conf()->mon_data);
```
Comment (athanatos) [inline src/ceph_mon.cc:359]: make_unique?

**Review by athanatos — APPROVED** [review]

File: src/ceph_mon.cc (hunk 2 above)
Comment (Matan-B) [reply src/ceph_mon.cc]: I don't see a reason not to, same for the Throttles. Updated.

**Review by batrick — APPROVED** [review]

**Comment by ljflores** [comment]
@Matan-B can you take a look at https://pulpito.ceph.com/lflores-2023-08-16_18:14:59-rados-wip-yuri4-testing-2023-08-10-1739-distro-default-smithi/7370221/?

This job failed twice in QA, and passed in a rebase w/o this PR. It wasn't obvious to me exactly what caused the failure, but the reason for failure was `expected valgrind issues and found none `.

Looking at the test, we intentionally leak memory in the mon, which should be detected by valgrind in the test. However, with this PR, the memory leak is not detected by valgrind when it should.

**Comment by ljflores** [comment]
Let me know when it's ready for testing again

**Review by rishabh-d-dave — APPROVED** [review]
CephFS integration tests ran successfully - https://tracker.ceph.com/projects/cephfs/wiki/Main#6-Sep-2023-Run-2

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcJxeRJYse1822-wip-yuri4-testing-2023-08-10-1739

Sorry for the major delay on this.

**Pull request merged**